### ISSUE TITLE
Trouble executing script with specified library location

### ISSUE THREAD

**Issue by frederikvand** [Circuitscape/Circuitscape.jl#214]
Dear,

Because I can't install Julia and Circuitscape in the default user folder on the supercomputer due to restrictions I changed all default locations with the following code:

JULIA_DEPOT_PATH= "/data/leuven/330/vsc33060/Julia/Packages"
JULIA_LOAD_PATH="/data/leuven/330/vsc33060/Julia/Packages"
ENV["JULIA_PKGDIR"] ="/data/leuven/330/vsc33060/Julia/Packages"
JULIA_PKGDIR="/data/leuven/330/vsc33060/Julia/Packages"
DEPOT_PATH[1] = "/data/leuven/330/vsc33060/Julia/Packages"
HOME = "/data/leuven/330/vsc33060/Julia/Packages"
ENV["HOME"] ="/data/leuven/330/vsc33060/Julia/Packages"
cd("/data/leuven/330/vsc33060/Julia/Packages")
using Pkg
Pkg.activate("/data/leuven/330/vsc33060/Julia/Packages/circuit")
Pkg.add("Circuitscape")
using Circuitscape
compute("/scratch/leuven/330/vsc33060/circuitscape/circuit/initest.ini")

However when I then compute the script it does not find circuitscape: 
This script did work partly before changing the path. Am I missing something? I also tried with 
julia> import Circuitscape
julia> Circuitscape.compute("/scratch/leuven/330/vsc33060/circuitscape/circuit/initest.ini") 
=> same result

[ Info: 2020-01-02 18:18:19 : Logs will recorded to file: log_file
[ Info: 2020-01-02 18:18:19 : Precision used: Double
[ Info: 2020-01-02 18:18:19 : Starting up Circuitscape to use 12 processes in parallel
ERROR: On worker 2:
ArgumentError: Package Circuitscape not found in current path:
- Run `import Pkg; Pkg.add("Circuitscape")` to install the Circuitscape package.

require at ./loading.jl:876
eval at ./boot.jl:330
top-level scope at none:0
eval at ./boot.jl:330
#116 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:307
run_work_thunk at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:79
run_work_thunk at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:88
#102 at ./task.jl:268
#remotecall_wait#154(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Distributed.remotecall_wait), ::Function, ::Distributed.Worker, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:421
remotecall_wait(::Function, ::Distributed.Worker, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:412
#remotecall_wait#157(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Distributed.remotecall_wait), ::Function, ::Int64, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:433
remotecall_wait at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:433 [inlined]
(::getfield(Distributed, Symbol("##161#163")){Module,Expr,Int64})() at ./task.jl:268

...and 59 more exception(s).

Stacktrace:
 [1] sync_end(::Array{Any,1}) at ./task.jl:235
 [2] macro expansion at ./task.jl:254 [inlined]
 [3] remotecall_eval(::Module, ::Array{Int64,1}, ::Expr) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/macros.jl:199
 [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/macros.jl:183 [inlined]
 [5] myaddprocs(::Int64) at /data/leuven/330/vsc33060/Julia/Packages/packages/Circuitscape/3Rn8u/src/utils.jl:127
 [6] compute(::String) at /data/leuven/330/vsc33060/Julia/Packages/packages/Circuitscape/3Rn8u/src/run.jl:29
 [7] top-level scope at REPL[41]:1

Thank you!

**Comment by ViralBShah** [comment]
@kristofferc Would you know what is the missing step here?

**Comment by KristofferC** [comment]
- `ENV["JULIA_PKGDIR"] ="/data/leuven/330/vsc33060/Julia/Packages"` can be removed, it is not used in 1.x
- `DEPOT_PATH[1] = "/data/leuven/330/vsc33060/Julia/Packages"` can be removed since you are already setting the `DEPOT_PATH` with `JULIA_DEPOT_PATH` (note that this envirnment variable has to be set before starting Julia)
- `JULIA_DEPOT_PATH= "/data/leuven/330/vsc33060/Julia/Packages"` could be changed to remove the last `Packages` directory since otherwise you will get e.g. a `Julia/Packages/packages` folder.
-  `JULIA_LOAD_PATH="/data/leuven/330/vsc33060/Julia/Packages"` should be removed. Instead you should start Julia with `--project=/path/to/project` which might be `--project=/data/leuven/330/vsc33060/Julia/` depending on where you have the Project.toml.

**Comment by ViralBShah** [comment]
Please reopen if the issue persists.

**Comment by frederikvand** [comment]
Thanks a lot for your feedback! The problem persists with the proposed changes although the Pkg.test("Circuitscape") is completed without issues.

CURRENT CODE : 

bash-4.2$ export JULIA_DEPOT_PATH="/data/leuven/330/vsc33060/Julia:$JULIA_DEPOT_PATH"

bash-4.2$ cd /scratch/leuven/330/vsc33060/circuitscape/circuit

bash-4.2$ module load Julia

bash-4.2$ julia --project=/data/leuven/330/vsc33060/Julia/circuit
julia> include("/data/leuven/330/vsc33060/input/Julia_scripts/circuit.jl")

circuit.jl 
JULIA_PKGDIR="/data/leuven/330/vsc33060/Julia"
HOME = "/data/leuven/330/vsc33060/Julia"
ENV["HOME"] ="/data/leuven/330/vsc33060/Julia"
cd("/data/leuven/330/vsc33060/Julia")

#first 4 lines don't make a difference if included or not

using Pkg
using Circuitscape
compute("/scratch/leuven/330/vsc33060/circuitscape/circuit/initest.ini")

#manual testing after script failure

julia> Pkg.test("Circuitscape")
[ Info: 2020-01-03 17:13:21 : Time taken to complete job = 15.261952334
[ Info: Test allToOneVerify12 passed
Test Summary:          | Pass  Total
Double Precision Tests |  578    578
   Testing Circuitscape tests passed 

julia> include("/data/leuven/330/vsc33060/input/Julia_scripts/circuit.jl")
[ Info: 2020-01-03 17:15:19 : Logs will recorded to file: log_file
[ Info: 2020-01-03 17:15:19 : Precision used: Double
[ Info: 2020-01-03 17:15:19 : Starting up Circuitscape to use 12 processes in parallel
ERROR: LoadError: On worker 2:
ArgumentError: Package Circuitscape not found in current path:
- Run `import Pkg; Pkg.add("Circuitscape")` to install the Circuitscape package.

julia> import Pkg

julia> Pkg.add("Circuitscape")
 Resolving package versions...
  Updating `/vsc-hard-mounts/leuven-data/330/vsc33060/Julia/circuit/Project.toml`
 [no changes]
  Updating `/vsc-hard-mounts/leuven-data/330/vsc33060/Julia/circuit/Manifest.toml`
 [no changes]

Error persists

The problem is definately related to the package installation location but I do not understand why Pkg.test succeeds while the julia script does not. The default location is in a user folder with a restriction of 10000 files and thus can not be used for package installation (probably why I got the write error before during processing). The julia main installation folder is not accessible for supercomputer users. A specific installation is thus the only solution.

**Comment by KristofferC** [comment]
>  [ Info: 2020-01-03 17:15:19 : Starting up Circuitscape to use 12 processes in parallel

Hm, it is something with that the new processes started with `addprocs` do not have the same things set as you.

Could you, as a test add

```
@everywhere Base.ACTIVE_PROJECT[] = Base.active_project()
```

above the line at

https://github.com/Circuitscape/Circuitscape.jl/blob/7d5d23c7cf179ccb68c7ccbef2d13139d61bb72d/src/utils.jl#L127

**Comment by frederikvand** [comment]
Thank you KristofferC,

The problem persists with the adapted utils.jl file. 
julia> import Circuitscape; pathof(Circuitscape)
"/data/leuven/330/vsc33060/Julia/packages/Circuitscape/3Rn8u/src/Circuitscape.jl"

julia> compute("/scratch/leuven/330/vsc33060/circuitscape/circuit/initest.ini")
[ Info: 2020-01-03 17:52:33 : Logs will recorded to file: log_file
[ Info: 2020-01-03 17:52:33 : Precision used: Double
[ Info: 2020-01-03 17:52:33 : Starting up Circuitscape to use 12 processes in parallel
ERROR: On worker 2:
ArgumentError: Package Circuitscape not found in current path:
- Run `import Pkg; Pkg.add("Circuitscape")` to install the Circuitscape package.

**Comment by KristofferC** [comment]
If you do

```
@everywhere println(Base.active_project())
@everywhere println(LOAD_PATH)
```

there, what does it show?

**Comment by frederikvand** [comment]
# Helps start new processes from the INI file
function myaddprocs(n)
    addprocs(n)
	@everywhere println(Base.active_project())
	@everywhere println(LOAD_PATH)
end

Same error when I add all 4 lines to the utils.jl file

julia> compute("/scratch/leuven/330/vsc33060/circuitscape/circuit/initest.ini")
[ Info: 2020-01-03 18:34:42 : Logs will recorded to file: log_file
[ Info: 2020-01-03 18:34:42 : Precision used: Double
[ Info: 2020-01-03 18:34:42 : Starting up Circuitscape to use 12 processes in parallel
ERROR: On worker 2:
ArgumentError: Package Circuitscape not found in current path:
- Run `import Pkg; Pkg.add("Circuitscape")` to install the Circuitscape package.

require at ./loading.jl:876
eval at ./boot.jl:330
top-level scope at none:0
eval at ./boot.jl:330
#116 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:307
run_work_thunk at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:79
run_work_thunk at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/process_messages.jl:88
#102 at ./task.jl:268
#remotecall_wait#154(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Distributed.remotecall_wait), ::Function, ::Distributed.Worker, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:421
remotecall_wait(::Function, ::Distributed.Worker, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:412
#remotecall_wait#157(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Distributed.remotecall_wait), ::Function, ::Int64, ::Module, ::Vararg{Any,N} where N) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:433
remotecall_wait at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/remotecall.jl:433 [inlined]
(::getfield(Distributed, Symbol("##161#163")){Module,Expr,Int64})() at ./task.jl:268

...and 95 more exception(s).

Stacktrace:
 [1] sync_end(::Array{Any,1}) at ./task.jl:235
 [2] macro expansion at ./task.jl:254 [inlined]
 [3] remotecall_eval(::Module, ::Array{Int64,1}, ::Expr) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/macros.jl:199
 [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Distributed/src/macros.jl:183 [inlined]
 [5] myaddprocs(::Int64) at /data/leuven/330/vsc33060/Julia/packages/Circuitscape/3Rn8u/src/utils.jl:127
 [6] compute(::String) at /data/leuven/330/vsc33060/Julia/packages/Circuitscape/3Rn8u/src/run.jl:29
 [7] top-level scope at REPL[38]:1

**Comment by frederikvand** [comment]
Currently there is only the following code in the Project.toml file 
[deps]
Circuitscape = "2b7a1792-8151-5239-925d-e2b8fdfa3201"

Anything I can add there so it actually finds the package? 

Thank you for all the help.

**Comment by KristofferC** [comment]
Yeah, I didn't expect the error to go away but that the printout could possibly diagnose it.

For some reason the workers started by `addprocs` can't find `Circuitscape` and I am not really sure why.

Could you try set

```
export JULIA_LOAD_PATH=/data/leuven/330/vsc33060/Julia/circuit
```

(before starting Julia and try again)

**Comment by frederikvand** [comment]
bash-4.2$ export JULIA_LOAD_PATH=/data/leuven/330/vsc33060/Julia/circuit
seems to have fixed initial problems after I also reinstalled and added Pkg to my project file!

The full script is running on the superdome cluster now.

There is an initial error /var/spool/torque/mom_priv/prologue: line 44: echo: write error: Invalid argument but the script is still running. This is the same error I got before but was never solved (in november) - error occurs in parallel and with 1 process. In the past the script only aborted after a couple of hours when it was writing output with a traceback on line 44 also. I will open a new ticket for that issue if the script aborts with traceback.

Thanks a lot for all the help!!!

**Comment by ViralBShah** [comment]
Thanks @KristofferC

**Comment by frederikvand** [comment]
Yes thank you @KristofferC!

**Comment by KristofferC** [comment]
No problem. Loading packages on workers is something that should be improved.
